Introduce the cypress/factory image (#812)
cypress-io/cypress-docker-images: factory/README.md

# cypress/factory

> cypress/factory is a docker container that can be used with `args` to generate a docker container with specific versions of:

Reviewer: IMO We should really spell out this is mean for our users, not necessarily for cypress-docs to build on.

* node

> * yarn

Reviewer: always both or could it be one or the other?

Author: always node, yarn is optional

* chrome
* firefox
* edge
* cypress

## Benefits

* Freedom to choose which versions to test against.
* No need to wait on an official release to test the latest version of a browser.
* Smaller docker sizes especially when not including unused browsers.
* Easily test multiple browser versions.
* Reduced maintenance and pull requests for the cypress-docker repo.
* Ability to offer more variations of docker containers at low cost.

## Usage

> The cypress docker factory works by relying on the `ONBUILD` docker command to run commands at build time of the the consuming container. `Args` can be specified in a number of ways, some of which are demonstrated below. For each of these examples we are building the equivalent of the `cypress/browsers` docker container. In each instance since the example is only testing the chrome version, the examples could just install chrome by itself if the other browsers were not used.

Reviewer: ```suggestion
The cypress docker factory works by relying on the [`ONBUILD`](https://docs.docker.com/engine/reference/builder/#onbuild) docker instruction to run commands at the container's build time. Docker `args` can be specified in a number of ways, some of which are demonstrated below. For each of these examples we are building the equivalent of the `cypress/browsers` docker image. In each instance since the example is only testing the chrome version, the examples could just install chrome by itself if the other browsers were not used.
```
> In each instance since the example is only testing the chrome version, the examples could just install chrome by itself if the other browsers were not used.

I'd be helpful to make this note later on and provide an example users can copy & paste and tweak.

### In the dockerfile

> Args can be defined directly in the dockerfile.

Reviewer: ```suggestion
Args can be defined directly in the dockerfile to specify variables that are available when the container is built.
```

You know alot about docker, a few filler details to provide context would be helpful in these examples for newbies.

dockerfile

```dockerfile

> # Args are defined in the dockerfile before the FROM command.

Reviewer: No node version was defined, can we call out the default?

# Using these args will cause an image to be created with node, chrome, firefox and edge.
ARG CHROME_VERSION='107.0.5304.121-1'
ARG EDGE_VERSION='100.0.1185.29-1'
ARG FIREFOX_VERSION='107.0'

FROM cypress/factory

COPY . /opt/app
WORKDIR /opt/app
RUN npm install --save-dev cypress
```

> build commands

Reviewer: ```suggestion
Then, is the same directory as the dockerfile, run the following commands to build the docker container and run Cypress against the chrome browser.
```

```bash
docker build . -t test
docker run -it test npm run test -b chrome
```

### At build time

> Args can be passed to the docker build command.

Reviewer: ```suggestion
Args can be passed to the docker build command with the `--build-arg` flag. Note: any value set via the command line will override the ARG value provided in the Dockerfile.
```

dockerfile

```dockerfile
FROM cypress/factory

COPY . /opt/app
WORKDIR /opt/app
RUN npm install --save-dev cypress
```

build commands

```bash
docker build . --build-arg CHROME_VERSION='107.0.5304.121-1' --build-arg EDGE_VERSION='100.0.1185.29-1' --build-arg FIREFOX_VERSION='107.0' -t test

docker run test npm run test -b chrome
```

### In docker-compose.yml

Finally, args can be specified in the docker-compose.yml file.

docker-compose.yml

```yml
version: '3'

services:
  test:
    args:
      CHROME_VERSION: '107.0.5304.121-1'
      EDGE_VERSION: '100.0.1185.29-1'
      FIREFOX_VERSION: '107.0'
    build:
      context: .
    command: npm run test
```

dockerfile

```dockerfile
FROM cypress/factory

COPY . /opt/app
WORKDIR /opt/app
RUN npm install --save-dev cypress
```

docker commands

```bash
docker-compose build test

docker-compose run test
```

> 

Reviewer: I'd add the example about only running chrome and you could build a slimmer browser image here

Author: added

> ## API

Reviewer: The API might be helpful to include before any examples IMO.

Author: I went back and forth on this, really i'm happy either way. I'll move it up.

The following args can be set to define what versions the cypress factory docker includes in it's final build.

If no args are defined, only the default version of node will be installed. This can still be a useful container though since we will also install any required cypress apt dependencies.

### NODE_VERSION

> The version of Node to install in the docker container. If not specified, the default version of Node (defined [here](./docker-compose.yml)) is installed. Node is required.

Reviewer: Should the link be to the .env file?  Also, will this typically match a cypress minimum version or how are we bumping this?

Author: oh, yes. it used to be defined in docker-compose but i moved it.

Author: It seems like it should typically match the bundled node version with cypress, but i'm not sure thats currently the case.

Reviewer: Matching the bundled version seems safe. Assume this means we'll likely release a MVB of the factory image when we release a new cypress major version?

Author: yeah, we'd have to

Author: if we wanted to keep the node version pegged.

Author: at a minimum i think we'd release a major version of factory if we drop support for the default node version.

Author: we could always build base/browsers/included with a different node version than the default.

Example: `NODE_VERSION='16.18.1'`

[Node Versions](https://nodejs.org/en/download/releases)

### YARN_VERSION

The version of yarn to install (via npm). If not specified, Yarn is not installed.

Example: `YARN_VERSION='1.22.19'`

[Yarn versions](https://www.npmjs.com/package/yarn)

### CYPRESS_VERSION

The version of Cypress to install (via npm). If not specified, Cypress is not installed.

Example: `CYPRESS_VERSION='12.1.0'`

[Cypress versions](https://www.npmjs.com/package/cypress)

> ### CHROME_VERSION

Reviewer: chrome, edge, firefox and yarn don't appear to be installed via npm. copy-paste err?

Author: yes

The version of Chrome to install (via npm). If not specified, Chrome is not installed.

Example: `CHROME_VERSION='107.0.5304.121-1'`

[Chrome versions](https://www.ubuntuupdates.org/package/google_chrome/stable/main/base/google-chrome-stable)

### FIREFOX_VERSION

The version of Firefox to install (via npm). If not specified, Firefox is not installed.

Example: `FIREFOX_VERSION='107.0'`

[Firefox versions](https://download-installer.cdn.mozilla.net/pub/firefox/releases/)

### EDGE_VERSION

The version of Edge to install (via npm). If not specified, Edge is not installed.

> Example: `EDGE_VERSION='100.0.1185.29-1'`

Reviewer: Do these have to be minor versions or can I do `EDGE_VERSION=100`?

Reviewer: I feel like someones gonna totally try this 😋

Author: Yes it deserves mentioning in the api doc

Author: updated

[Edge versions](https://packages.microsoft.com/repos/edge/pool/main/m/microsoft-edge-stable/)

## Version Testing

> Due to the large amount of possible version combinations, we're not able to exhaustively test each combination of versions, nor do we block versions that are incompatible. For example, Cypress 12 removed support for node versions less than 16.16.0. You are still able to generate a container with node 16.0.0 and Cypress 12, but Cypress will fail to run. This is because the factory supports earlier versions of Cypress and must support earlier versions of node.

Reviewer: ```suggestion
Due to the large amount of possible version combinations, we're not able to exhaustively test each combination of versions, nor do we block versions that are incompatible. For example, Cypress 12 removed support for node versions less than 16.0.0. You are still able to generate a container with node 14.0.0 and Cypress 12, but Cypress will fail to run. This is because the factory supports earlier versions of Cypress and must support earlier versions of node.
```

If you run across a combination that should reasonably work, but doesn't, log an issue and we will take a look at supporting it.

> Additionally this docker container and containers generated from it are intended for test use only, and are not intended for use hosting services in a production environment.

Reviewer: ```suggestion
Additionally this docker image and containers generated from it are intended for test use only, and are not intended for use hosting services in a production environment.
```

# cypress/factory

`cypress/factory` is a docker image that can be used with docker [`args`](https://docs.docker.com/engine/reference/builder/#arg) to generate a docker container with specific versions of:

* node
* yarn
* chrome
* firefox
* edge
* cypress

## Benefits

* Freedom to choose which versions to test against.
* No need to wait on an official release to test the latest version of a browser.
* Smaller docker sizes especially when not including unused browsers.
* Easily test multiple browser versions.
* Reduced maintenance and pull requests for the cypress-docker repo.

> * Ability to offer more variations of docker containers at low cost.

Reviewer 2: Just for my edification, what is the cost incurred here (which is potentially now reduced)?

Author: This is a potential cost savings. If we want to offer say a container that contains just cypress and node, previously we'd have to write a new script, make new folders and manually update each version. With the new CI and .env setup all we have to do is update CI and a couple of docker-compose files and new versions would be released automatically in sync with the `included` image based on the version of cypress and node in the .env file.

Reviewer 2: That's what I assumed it implied. It's really an internal cost/time savings. Its minor, but may be worth clarifying (I think it speaks loudly to a fundamental impact here), to something like _Ability for Cypress to offer_ so its a known savings commodity that we'd be paying forward.

Author: ```suggestion
* Ability for Cypress to offer more variations of docker containers at low cost.
```

## API

> The following args can be set to define what versions the cypress factory docker includes in it's final build.

Reviewer 2: Maybe a small update like this to clarify:

```suggestion
The following args can be set to define what versions the cypress factory includes in its final build.
```

or

```suggestion
The following args can be set to define what versions the cypress factory image includes in its final build.
```

> If no args are defined, only the default version of node will be installed. This can still be a useful container though since we will also install any required cypress apt dependencies.

Reviewer 2: Can you clarify this line? Is it _cypress app dependencies_?

Reviewer 2: dig it!

### NODE_VERSION

The version of Node to install in the docker container. If not specified, the default version of Node (defined [here](./.env)) is installed. Node is required. The exact version must be used, no wildcards or shorthands are supported.

Example: `NODE_VERSION='16.18.1'`

[Node Versions](https://nodejs.org/en/download/releases)

### YARN_VERSION

> The version of yarn to install (via npm). If not specified, Yarn is not installed.

Author: I'll update these descriptions

> The version of yarn to install (via npm). If not specified, Yarn is not installed.

Author: updated

Example: `YARN_VERSION='1.22.19'`

[Yarn versions](https://www.npmjs.com/package/yarn)

### CYPRESS_VERSION

> The version of Cypress to install (via npm). If not specified, Cypress is not installed.

Author: updated

Example: `CYPRESS_VERSION='12.1.0'`

[Cypress versions](https://www.npmjs.com/package/cypress)

### CHROME_VERSION

The version of Chrome to install. If not specified, Chrome is not installed. The exact version must be used, no wildcards or shorthands are supported.

Example: `CHROME_VERSION='107.0.5304.121-1'`

[Chrome versions](https://www.ubuntuupdates.org/package/google_chrome/stable/main/base/google-chrome-stable)

### FIREFOX_VERSION

The version of Firefox to install. If not specified, Firefox is not installed. The exact version must be used, no wildcards or shorthands are supported.

Example: `FIREFOX_VERSION='107.0'`

[Firefox versions](https://download-installer.cdn.mozilla.net/pub/firefox/releases/)

### EDGE_VERSION

The version of Edge to install. If not specified, Edge is not installed. The exact version must be used, no wildcards or shorthands are supported.

Example: `EDGE_VERSION='100.0.1185.29-1'`

[Edge versions](https://packages.microsoft.com/repos/edge/pool/main/m/microsoft-edge-stable/)

> ## Usage

Reviewer 3: I guess all this assumes that there is already a `cypress/factory` image in the user's registry. How would a user build and tag `cypress/factory` locally? I tried `docker build . -f ./factory.Dockerfile -t cypress/factory` but got:

```
Step 1/21 : ARG BASE_IMAGE
Step 2/21 : FROM ${BASE_IMAGE} as factory
base name (${BASE_IMAGE}) should not be blank
```

...so I think I'm missing something. Could go at the end of this `Usage` section.

Author: We'll be publishing cypress/factory to dockerhub so people will have access to it. I don't really expect people to build it themselves.

for us we can run the following from the inside the `factory` directory.

`docker-compose build factory`

Reviewer 3: It would be good to add that to this doc for future contributors. And also for me, because I didn't know how to test this locally :smile:

Author: Thats a good call.

Reviewer 3: That command doesn't seem to work for me:

```
➜  factory git:(matth/docker-factory) docker-compose build factory
ERROR: The Compose file './docker-compose.yml' is invalid because:
services.factory.build contains unsupported option: 'tags'
services.chrome.build contains unsupported option: 'tags'
services.cypress-chrome.build contains unsupported option: 'tags'
services.cypress-edge.build contains unsupported option: 'tags'
services.cypress-firefox.build contains unsupported option: 'tags'
services.cypress.build contains unsupported option: 'tags'
services.edge.build contains unsupported option: 'tags'
services.firefox.build contains unsupported option: 'tags'
services.browsers.build contains unsupported option: 'tags'
services.base.build contains unsupported option: 'tags'
services.included.build contains unsupported option: 'tags'
```

```
➜  factory git:(matth/docker-factory) docker --version
Docker version 20.10.22, build 3a2c30b
➜  factory git:(matth/docker-factory) docker-compose --version
docker-compose version 1.29.2, build unknown
```

It says my Docker version should support `3.8` `docker-compose` files, any idea why it's not working?

Author: I had to update to the latest docker-compose version to get docker-compose to support the `tags` option

```
➜  factory git:(matth/docker-factory) docker --version
Docker version 20.10.21, build baeda1f
➜  factory git:(matth/docker-factory) docker-compose --version
Docker Compose version v2.13.0
```

Reviewer 3: Thanks. Weird... Docker's default Ubuntu install has `docker-compose` 1.x, but if you install `docker-compose-plugin`, you can use `docker compose` 2.x...

```
➜  ~ docker compose version  
Docker Compose version v2.14.1
➜  ~ docker-compose --version
docker-compose version 1.29.2, build unknown
```

So with `docker compose` I am able to run that command to build the factory image.

Might also be worth mentioning that `docker-compose` 2.x is necessary because for some reason it seems to be elusive.

Author: thats good to know. I might go through and update any `docker-compose` commands to `docker compose` instead.

Author: Added instructions to build and test locally in the contributing document: https://github.com/cypress-io/cypress-docker-images/pull/812/commits/60088df3be182cee846520b9eb7be23de853d6f7

> The cypress docker factory works by relying on the [`ONBUILD`](https://docs.docker.com/engine/reference/builder/#onbuild) docker instruction to run commands at the container's build time. Docker `args` can be specified in a number of ways, some of which are demonstrated below. For each of these examples we are building the equivalent of the `cypress/browsers` docker image. In each instance since the example is only testing the chrome version, the examples could just install chrome by itself if the other browsers were not used.

Author: I updated the description the specify that users would need to create a dockerfile.

### In the Dockerfile

Args can be defined directly in the Dockerfile to specify variables that are available when the container is built.

Dockerfile

```dockerfile
# Args are defined in the Dockerfile before the FROM command.
# Using these args will cause an image to be created with node (default version is 16.18.1), chrome, firefox and edge.
ARG CHROME_VERSION='107.0.5304.121-1'
ARG EDGE_VERSION='100.0.1185.29-1'
ARG FIREFOX_VERSION='107.0'

FROM cypress/factory

COPY . /opt/app
WORKDIR /opt/app
RUN npm install --save-dev cypress
```

> Then, is the same directory as the Dockerfile, run the following commands to build the docker container and run Cypress against the chrome browser.

Reviewer 2: Should this read _in_ the same directory?

Author: yes
```suggestion
Then, in the same directory as the Dockerfile, run the following commands to build the docker container and run Cypress against the chrome browser.
```

docker run test npm run test -b chrome
```

### In docker-compose.yml

Finally, args can be specified in the docker-compose.yml file.

docker-compose.yml

```yml
version: '3'

services:
  test:
    args:
      CHROME_VERSION: '107.0.5304.121-1'
      EDGE_VERSION: '100.0.1185.29-1'
      FIREFOX_VERSION: '107.0'
    build:
      context: .
    command: npm run test
```

Dockerfile

```dockerfile
FROM cypress/factory

COPY . /opt/app
WORKDIR /opt/app
RUN npm install --save-dev cypress
```

docker commands

```bash
docker compose build test

docker compose run test
```

### Reducing the size of the docker container

As mentioned above we can reduce the size of the docker image by removing browsers we aren't using.

Since this example only uses chrome, removing edge and firefox is as simple as not including a version.

Dockerfile

```dockerfile
# Args are defined in the Dockerfile before the FROM command.
# Using these args will cause an image to be created with node (default version is 16.18.1), chrome, firefox and edge.
ARG CHROME_VERSION='107.0.5304.121-1'

FROM cypress/factory

COPY . /opt/app
WORKDIR /opt/app
RUN npm install --save-dev cypress
```

> Then, is the same directory as the Dockerfile, run the following commands to build the docker container and run Cypress against the chrome browser.

Reviewer 2: same here _in_?

Author: yep!
```suggestion
Then, in the same directory as the Dockerfile, run the following commands to build the docker container and run Cypress against the chrome browser.
```

```bash
docker build . -t test
docker run -it test npm run test -b chrome
```

The docker image including chrome, edge and firefox weighs in a ~1.93 GB, by removing edge and firefox the image can be reduced to ~1.06 GB.

## Version Testing

Due to the large amount of possible version combinations, we're not able to exhaustively test each combination of versions, nor do we block versions that are incompatible. For example, Cypress 12 removed support for Node.js version 12.0.0. You are still able to generate a container with node 12.0.0 and Cypress 12, but Cypress will fail to run. This is because the factory supports earlier versions of Cypress and must support earlier versions of node.

If you run across a combination that should reasonably work, but doesn't, log an issue and we will take a look at supporting it.

> Additionally this docker image and containers generated from it are intended for test use only, and are not intended for use hosting services in a production environment.

Author: ```suggestion
Additionally this docker image and containers generated from it are intended for test use only, and are not intended for hosting services in a production environment.
```

docker-compose run test
```

### Reducing the size of the docker container

> As mentioned above we can reduce the size of the docker image by removing browsers we are using.

Reviewer: ```suggestion
As mentioned above we can reduce the size of the docker image by removing browsers we aren't using.
```

Author: Oops!
